src/fireedge/src/server/routes/api/oneflow/service/functions.js

```diff
@@ -0,0 +1,678 @@
+    const v = new Validator()
+    const postAction = parsePostData(serviceAction)
+    const valSchema = v.validate(postAction, action)
+    if (valSchema.valid) {
+      // validate if "action" is required
+      const config = {
+        method: POST,
+        path: '/service/{0}/role_action',
+        user,
+        password,
+        request: id,
+        post: { action: postAction },
+      }
+      oneFlowConnection(
+        config,
+        (data) => success(next, res, data),
+        (data) => error(next, res, data)
+      )
+    } else {
+      res.locals.httpCode = httpResponse(
+        internalServerError,
+        '',
+        `invalid schema ${returnSchemaError(valSchema.errors)}`
+      )
+      next()
+    }
+  } else {
+    res.locals.httpCode = httpResponse(
+      methodNotAllowed,
+      '',
+      'invalid action, id service or role'
+    )
+    next()
+  }
+}
+
+/**
+ * Add service role action.
+ *
+ * @param {object} res - http response
+ * @param {Function} next - express stepper
+ * @param {object} params - params
+ * @param {number} params.id - service ID
+ * @param {string} params.action - service action
+ * @param {string} params.role - service role
+ * @param {object} userData - user data
+ * @param {string} userData.user - username
+ * @param {string} userData.password - user password
+ */
+const serviceAddRoleAction = (
+  res = {},
+  next = defaultEmptyFunction,
+  params = {},
+  userData = {}
+) => {
+  const { user, password } = userData
+  const { role, id, action: serviceAction } = params
+  if (
+    role &&
+    Number.isInteger(parseInt(id, 10)) &&
+    serviceAction &&
+    user &&
+    password
+  ) {
+    const v = new Validator()
+    const postAction = parsePostData(serviceAction)
+    const valSchema = v.validate(postAction, action)
+    if (valSchema.valid) {
+      // validate if "action" is required
+      const config = {
+        method: POST,
+        path: '/service/{0}/role/{1}/action',
+        user,
+        password,
+        request: [id, role],
+        post: { action: postAction },
+      }
+      oneFlowConnection(
+        config,
+        (data) => success(next, res, data),
+        (data) => error(next, res, data)
+      )
+    } else {
+      res.locals.httpCode = httpResponse(
+        internalServerError,
+        '',
+        `invalid schema ${returnSchemaError(valSchema.errors)}`
+      )
+      next()
+    }
+  } else {
+    res.locals.httpCode = httpResponse(
+      methodNotAllowed,
+      '',
+      'invalid action, id service or role'
+    )
+    next()
+  }
+}
+
+/**
+ * Get service info.
+ *
+ * @param {string} user - username
+ * @param {string} password - password
+ * @param {string} serviceID - service ID
+ * @param {Function} succss - callback when have service info data
+ * @param {Function} error - callback when no have service info data
+ */
+const getNodesService = (
+  user = '',
+  password = '',
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Avoid using `==` and `===` operators with sensitive data fields like passwords, api keys, tokens etc. as they are vulnerable to timing attack. Consider using `crypto.timingSafeEqual` or `secure-compare` for time-safe string comparison.

[Learn more](https://cwe.mitre.org/data/definitions/208.html)